src/file_watching.jl

```diff
@@ -1,56 +1,153 @@
 end
 
 """
-    _file_watcher(filepath, fw)
+    _file_watcher(w::SimpleWatcher)
 
-Helper function that's called asynchronously to act if a file has changed.
-See also [`add_to_filewatcher`](@ref).
+[INTERNAL] Helper function that's spawned as an asynchronous task which
+checks for changed files. Terminates normally upon a `InterruptException`,
+and with a warning for all other exceptions.
 """
-function _file_watcher(filepath::AbstractString, fw::FileWatcher)
-    fileevent = watch_file(filepath, -1) # returns upon a change
-    # if file changed, start new task on this file (s.t. it is still being watched)
-    if !fileevent.renamed && !fileevent.timedout
-        fw.watched_files[filepath] = @async _file_watcher(filepath, fw)
-    else
-        # if renamed or timed out, remove from list of watched files
-        delete!(fw.watched_files, filepath)
+function _file_watcher(w::SimpleWatcher)
+    try
+        while true
+            # checking for changed files
+            changed_files = Vector{String}()
+            foreach(w.filelist) do wf
+                if has_changed(wf)
+                    set_unchanged!(wf)
+                    push!(changed_files, wf.path)
+                end
+            end
+
+            # invoke callback on each changed file (if it's defined), then sleep
+            (w.callback != nothing) && foreach(w.callback, changed_files)
+            sleep(w.sleeptime)
+        end
+    catch EXC
+        if !isa(EXC, InterruptException) # if interruption, normal termination
+            @warn "Exception in hdlr: " EXC
+        end
     end
-    # fire user-specified callback
-    fw.callback(filepath, fileevent)
 end
 
 """
-    add_to_filewatcher!(filewatcher, filepath)
+    _waitfor_task_shutdown(w::SimpleWatcher)
+
+[INTERNAL] Helper function ensuring that the `_file_watcher` task has ended
+before continuing.
+"""
+_waitfor_task_shutdown(w::SimpleWatcher) = while !istaskdone(w.task) sleep(0.05) end
+
+"""
+    set_callback(w::SimpleWatcher, fcn::Function)
+
+API function to set or change the callback function being executed upon a
+file change. Can be "hot-swapped", i.e. while the file watcher is running.
+The callback function receives a string with the file path and is not
+expected to return anything.
+"""
+function set_callback(w::SimpleWatcher, fcn::Function)
+    prev_running = stop(w) # returns true if was running
+    w.callback = fcn
+    prev_running && start(w) # start again if it was running before
+end
+
+"""
+    isrunning(w::SimpleWatcher)
+
+API function to check whether the file watcher is running. Should not be
+required though, since the `start` and `stop` API functions work in any
+case.
+"""
+isrunning(w::SimpleWatcher) = (w.task != nothing) && !istaskdone(w.task)
+
+"""
+    start(w::SimpleWatcher)
+
+API function to start the file watcher.
+"""
+start(w::SimpleWatcher) = !isrunning(w) && (w.task = @async _file_watcher(w))
```

Comment:
are we sure all of those get stopped?

Reply:
Yes. There can only be one at a time (see the `isrunning` function). And that one gets killed by the `stop` function which is called upon server termination.